Improved GPT4All model streaming tokens without adding callbacks

Anindyadeep opened the pull request:
- **Description:** Right now token streaming is enabled by adding callbacks. But there was no streaming of tokens occuring after adding callbacks.Latest version of gpt4all: version `1.0.3` addresses this issue and hence I tried to integrate that in the existing function.I have kept the code for supporting callbacks for streaming,  however I think that can be removed in coming versions. Also added docs for the same.
  
  - **Issue:** This commit fixes issue #7747.,
  - Dependencies: No dependency change,
  - Tag maintainer: @hwchase17 @agola11,
  - Twitter handle: @AnindyadeepS

Commit:
Improved GPT4All model streaming tokens without adding callbacks

This commit fixes issue #7747. Right now token streaming is enabled
by adding callbacks. But there was no streaming of tokens occuring
after adding callbacks.

Latest version of gpt4all: version `1.0.3` addresses this issue and hence
I tried to integrate that in the existing function.

I have kept the code for supporting callbacks for streaming,  however I think
that can be removed in coming versions.

Anindyadeep commented:
@hwchase17 @baskaryan Can you please review this PR and help me to understand why the lint checks for python 3.8 failed?

hinthornw commented on langchain/llms/gpt4all.py:
```diff
@@ -1,7 +1,9 @@
 from functools import partial
 from typing import Any, Dict, List, Mapping, Optional, Set
 
+import pkg_resources
```
I think we should opt for a different stdlib package if possible https://discuss.python.org/t/will-setuptools-remove-pkg-resource-module-in-the-future/27182

Anindyadeep replied:
@hinthornw We can do it in another way, the streaming for gpt4all throws `NotImplementedError` if used the previous version, hence we can detect there by a `try-except` block. In that way we can do it without `pkg_resource` 

Another thing, should we remove the callback support for streaming and make this changes as the default one? In that way there might be less ambiguity?

Anindyadeep replied:
@hinthornw I also wonder, Does providing custom StreamingCallback provide any kind of added advantage?

Commit:
This commit, I tried delete the clutter and just adding support for
streaming by just streaming=True/False. This makes it easy for developers
without using further callbacks.

Updated the docs and also it passes tests and lint checks

Anindyadeep commented:
@hinthornw can you please review this and provide your thoughts
Thanks

Commit:
Merge branch 'master' into anindya/gpt4all_streaming_fix

Anindyadeep commented:
@hinthornw , I am closing this PR, as I see now that adding this support explicitely might break the overall flow of how langchain's LLM wrapper works. As every wrapper is following a convention, along with how it will stream. So should I close this PR?
